Readme.md
# ![Retext](https://cdn.rawgit.com/wooorm/retext/master/logo.svg)

[![Build Status](https://img.shields.io/travis/wooorm/retext.svg?style=flat)](https://travis-ci.org/wooorm/retext) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext.svg?style=flat)](https://coveralls.io/r/wooorm/retext?branch=master) [![Code Climate](http://img.shields.io/codeclimate/github/wooorm/retext.svg?style=flat)](https://codeclimate.com/github/wooorm/retext)

**retext** is an extensible natural language system—by default using **[parse-latin](https://github.com/wooorm/parse-latin)** to transform natural language into a **[TextOM](https://github.com/wooorm/textom/)** object model. **Retext** provides a pluggable system for analysing and manipulating natural language in JavaScript. NodeJS and the browser. Tests provide 100% coverage.

> Rather than being a do-all library for Natural Language Processing (such as [NLTK](http://www.nltk.org) or [OpenNLP](https://opennlp.apache.org)), **retext** aims to be useful for more practical use cases (such as censoring profane words or decoding emoticons, but the possibilities are endless) instead of more academic goals (research purposes).
> **retext** is inherently modular—it uses plugins (similar to [rework](https://github.com/reworkcss/rework/) for CSS) instead of providing everything out of the box (such as [Natural](https://github.com/NaturalNode/natural)). This makes **retext** a viable tool for use on the web.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext
```

[Duo](http://duojs.org/#getting-started):

```javascript
var Retext = require('wooorm/retext');
```

UMD (globals/AMD/CommonJS) ([uncompressed](retext.js) and [compressed](retext.min.js)):

```html
<script src="path/to/retext.js"></script>
<script>
  var retext = new Retext();
</script>
```

## Usage

The following example uses **[retext-emoji](https://github.com/wooorm/retext-emoji)** (to show emoji) and **[retext-smartypants](https://github.com/wooorm/retext-smartypants)** (for smart punctuation).

```javascript
/* Require dependencies. */
var Retext = require('retext'),
    emoji = require('retext-emoji'),
    smartypants = require('retext-smartypants');

/* Create an instance using retext-emoji and -smartypants. */
var retext = new Retext()
    .use(emoji, {
        'convert' : 'encode'
    })
    .use(smartypants);

/* Read a document. */
retext.parse(
    'The three wise monkeys [. . .] sometimes called the ' +
    'three mystic apes--are a pictorial maxim. Together ' +
    'they embody the proverbial principle to ("see no evil, ' +
    'hear no evil, speak no evil"). The three monkeys are ' +
    'Mizaru (:see_no_evil:), covering his eyes, who sees no ' +
    'evil; Kikazaru (:hear_no_evil:), covering his ears, ' +
    'who hears no evil; and Iwazaru (:speak_no_evil:), ' +
    'covering his mouth, who speaks no evil.',
    function (err, tree) {
        /* Handle errors. */
        if (err) {
            throw err;
        }
        
        /* Log the text content of the tree (the transformed input). */
        console.log(tree.toString());
        /**
         * This logs the following:
         *   The three wise monkeys […] sometimes called the three
         *   mystic apes—are a pictorial maxim. Together they
         *   embody the proverbial principle to (“see no evil,
         *   hear no evil, speak no evil”). The three monkeys are
         *   Mizaru (🙈), covering his eyes, who sees no evil;
         *   Kikazaru (🙉), covering his ears, who hears no evil;
         *   and Iwazaru (🙊), covering his mouth, who speaks no evil.
         */
    }
);
```

## API

### Retext(parser?)

```javascript
var Retext = require('retext'),
    ParseEnglish = require('parse-english');

var retext = new Retext(new ParseEnglish());

/* There, ol’ chap. */
retext.parse('Some English', function (err, tree) {/* ... */});
```

Return a new `Retext` instance with the given [parser](#parsers) (defaults to an instance of **parse-latin**).

### Retext#use([plugin](#plugin), options?)

Takes a plugin—a humble function to transform the object model.
Optionally takes an `options` object, but it’s up to plugin authors to support settings.

### Retext#parse(value, options?, done(err, tree))

Parses the given source and, when done, passes either an error (the first argument), or the (by `use`d plugins, modified) document (the second argument) to the callback.

### plugin

A plugin is simply a function, with `function(retext, options?)` as its signature. The first argument is the **Retext** instance a user attached the plugin to. The plugin is invoked when a user `use`s the plugin (not when a document is parsed) and enables the plugin to modify the internal Object Model ([`retext.TextOM`](https://github.com/wooorm/textom)) or the parser ([`retext.parser`](https://github.com/wooorm/parse-latin)).

The plugin can return another function: `function(NLCSTNode, options, next?)`. This function is invokeded when a document is parsed. It’s given the document as created by `Retext#parse()` before it’s given to the user.

## Plugins

- [retext-content](https://github.com/wooorm/retext-content) — Append, prepend, remove, and replace content into/from Retext nodes;
- [retext-cst](https://github.com/wooorm/retext-cst) — (**[demo](http://wooorm.github.io/retext-cst/)**) — Encoding and decoding between AST (JSON) and TextOM object model;
- [retext-directionality](https://github.com/wooorm/retext-directionality) — (**[demo](http://wooorm.github.io/retext-directionality/)**) — Detect the direction text is written in;
- [retext-dom](https://github.com/wooorm/retext-dom) — (**[demo](http://wooorm.github.io/retext-dom/)**) — Create a (living) DOM tree from a TextOM tree;
- [retext-double-metaphone](https://github.com/wooorm/retext-double-metaphone) — (**[demo](http://wooorm.github.io/retext-double-metaphone/)**) — Implementation of the Double Metaphone algorithm;
- [retext-emoji](https://github.com/wooorm/retext-emoji) — (**[demo](http://wooorm.github.io/retext-emoji/)**) — Encode or decode [Gemojis](https://github.com/github/gemoji);
- [retext-find](https://github.com/wooorm/retext-find) — Easily find nodes;
- [retext-inspect](https://github.com/wooorm/retext-inspect) — (**[demo](http://wooorm.github.io/retext-inspect/)**) — Nicely display nodes in `console.log` calls;
- [retext-keywords](https://github.com/wooorm/retext-keywords) — (**[demo](http://wooorm.github.io/retext-keywords/)**) — Extract keywords and keyphrases;
- [retext-lancaster-stemmer](https://github.com/wooorm/retext-lancaster-stemmer) — (**[demo](http://wooorm.github.io/retext-lancaster-stemmer/)**) — Implementation of [the Lancaster (Paice/Husk) algorithm](http://www.comp.lancs.ac.uk/computing/research/stemming/index.htm);
- [retext-language](https://github.com/wooorm/retext-language) — (**[demo](http://wooorm.github.io/retext-language/)**) — Detect the language of text;
- [retext-link](https://github.com/wooorm/retext-link) — (**[demo](http://wooorm.github.io/retext-link/)**) — Detect links in text;
- [retext-live](https://github.com/wooorm/retext-live) — Change a node based on a (new?) value;
- [retext-metaphone](https://github.com/wooorm/retext-metaphone) — (**[demo](http://wooorm.github.io/retext-metaphone/)**) — Implementation of the Metaphone algorithm;
- [retext-porter-stemmer](https://github.com/wooorm/retext-porter-stemmer) — (**[demo](http://wooorm.github.io/retext-porter-stemmer/)**) — Implementation of [the Porter stemming algorithm](http://tartarus.org/martin/PorterStemmer/);
- [retext-pos](https://github.com/wooorm/retext-pos) — (**[demo](http://wooorm.github.io/retext-pos/)**) — Part-of-speech tagger;
- [retext-range](https://github.com/wooorm/retext-range) — Sequences of content within a TextOM tree between two points;
- [retext-search](https://github.com/wooorm/retext-search) — (**[demo](http://wooorm.github.io/retext-search/)**) — Search in a TextOM tree;
- [retext-sentiment](https://github.com/wooorm/retext-sentiment) — (**[demo](http://wooorm.github.io/retext-sentiment/)**) — Detect sentiment in text;
- [retext-smartypants](https://github.com/wooorm/retext-smartypants) — (**[demo](http://wooorm.github.io/retext-smartypants/)**) — Implementation of [SmartyPants](http://daringfireball.net/projects/smartypants/);
- [retext-soundex](https://github.com/wooorm/retext-soundex) — (**[demo](http://wooorm.github.io/retext-soundex/)**) — Implementation of the Soundex algorithm;
- [retext-syllable](https://github.com/wooorm/retext-syllable) — (**[demo](http://wooorm.github.io/retext-syllable/)**) — Syllable count;
- [retext-visit](https://github.com/wooorm/retext-visit) — (**[demo](http://wooorm.github.io/retext-visit/)**) — Visit nodes, optionally by type;
- [retext-walk](https://github.com/wooorm/retext-walk) — Walk trees, optionally by type.

## Desired Plugins

> Hey! Want to create one of the following, or any other plugin, for **retext** but not sure where to start? I suggest to read **retext-visit**’s source code to see how it’s build first (it’s probably the most straight forward to learn), and go from there.
> Let me know if you still have any questions, go ahead and send me [feedback](mailto:tituswormer@gmail.com) or [raise an issue](https://github.com/wooorm/retext/issues).

- retext-date — Detect time and date in text;
- retext-frequent-words — Like **retext-keywords**, but based on frequency and stop-words instead of a POS-tagger;
- retext-hyphen — Insert soft-hyphens where needed; this might have to be implemented with some sort of node which doesn’t stringify;
- retext-location — Track the position of nodes (line, column);
- retext-no-pants — Opposite of **retext-smartypants**;
- retext-no-break — Inserts [non-breaking spaces](http://en.wikipedia.org/wiki/Non-breaking_space#Non-breaking_behavior) between things like “100 km”;
- retext-profanity — Censor profane words;
- retext-punctuation-pair — Detect which opening or initial punctuation, belongs to which closing or final punctuation mark (and vice versa);
- retext-summary — Summarise text;
- retext-sync — Detect changes in a textarea (or contenteditable?), sync the diffs over to a **retext** tree, let plugins modify the content, and sync the diffs back to the textarea;
- retext-typography — Applies typographic enhancements, like (or using?) retext-smartypants and retext-hyphen;
- retraverse — Like Estraverse.

## Parsers

- [parse-latin](https://github.com/wooorm/parse-latin) (**[demo](http://wooorm.github.io/parse-latin/)**) — default;
- [parse-english](https://github.com/wooorm/parse-english) (**[demo](http://wooorm.github.io/parse-english/)**) — Specifically for English;
- [parse-dutch](https://github.com/wooorm/parse-dutch) (**[demo](http://wooorm.github.io/parse-dutch/)**) — Specifically for Dutch;

## Benchmark

On a MacBook Air, it parses about 2 big articles, 25 sections, or 230 paragraphs per second.

```text
           retext.parse(value, callback);
  230 op/s » A paragraph (5 sentences, 100 words)
   25 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
    2 op/s » An article (100 paragraphs, 500 sentences, 10,000 words)
```

## Related

- [textom](https://github.com/wooorm/textom)
- [nlcst](https://github.com/wooorm/nlcst)

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

n.n.n / 2015-01-23
==================

  * Add UMD as an installation method in `Readme.md`
  * Add `index.js` to bower ignore
  * Remove bower dependencies due to UMD build
  * Add `retext.js` to bowers `main` instead of `index.js`
  * Add `retext.js`, `retext.min.js`
  * Add `retext.js`, `retext.min.js` to `.npmignore`
  * Add `postbuild` npm script target to mangle build
  * Add `build`, `prepublish` npm script target
  * Add esmangle as a dev-dependency
  * Add browserify as a dev-dependency
  * Refactor npm script targets
  * Update eslint
  * Add demo for `retext-keywords` to `Readme.md`
  * Add demo for `retext-pos` to `Readme.md`

0.5.0 / 2015-01-12
==================

  * Update textom to a stable version
  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor `package.json`
  * Update jscs-jsdoc
  * Refactor `test.js` for changes in eslint
  * Update eslint

0.5.0-rc.2 / 2014-12-19
==================

  * Update textom
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Add demo for `retext-inspect` to `Readme.md`
  * Add demo for `retext-cst` to `Readme.md`

0.5.0-rc.1 / 2014-12-05
==================

 * Fix benchmark results in `Readme.md`
 * Fix typos in `Readme.md`
 * Fix example to be valid JS in `Readme.md`
 * Add link to personal website to copyright in `Readme.md`
 * Refactor npm script targets in `package.json`
 * Update benchmark results in `Readme.md`
 * Update nlcst-to-textom, textom
 * Update eslint
 * Remove retext-emoticon from desired plug-ins in `Readme.md`
 * Add demo for `retext-syllabe` to `Readme.md`
 * Rename retext-ast to retext-cst in plugins in `Readme.md`
 * Fix typo’s in `Readme.md`

0.4.0 / 2014-11-20
==================

 * Update dependencies
 * Fix incorrect executive rights on `test.js`

0.4.0-rc.2 / 2014-11-19
==================

 * Update parse-latin, textom to lates development releases
 * Fix logo link in `Readme.md`

0.4.0-rc.1 / 2014-11-15
==================

 * Refactor script targets in `package.json`
 * Add `logo.svg`
 * Add description of plugins to `Readme.md`
 * Remove personal intro from `Readme.md`
 * Add flat badges to `Readme.md`
 * Update parse-latin
 * Update matcha, nlcst-to-textom

0.3.0 / 2014-10-28
==================

 * Refactor example in `Readme.md`
 * Update textom, parse-latin

0.3.0-rc.2 / 2014-10-27
==================

 * Remove superfluous test code
 * Refactor docs in `Readme.md`
 * Refactor module
 * Remove `retext` argument for plugin's `onrun`
 * Merge branch 'feature/standardise-options'
 * Add functionality for options for `use` and `parse`
 * Add tests for standardised options for `use` and `parse`
 * Merge branch 'feature/refactor-plugin-and-attach-system'
 * Refactor module for new plugin/attach system
 * Refactor test for new plugin/attach system
 * Add `.eslintrc`
 * Disallow space after object keys in `.jscs.json`
 * Update eslint, mocha
 * Add retext-typography to desired plugins in Readme.md
 * Add retext-hyphen to desired plugins in Readme.md
 * Update retext-live (to -sync) in desired plugins in `Readme.md`
 * Add retext-live to plugins in Readme.md

0.3.0-rc.1 / 2014-10-21
==================

 * Update textom, parse-latin
 * Add retext-no-break to desired plugins in Readme.md
 * Add retext-location to desired plugins in Readme.md

0.2.1 / 2014-10-15
==================

 * Add test for `TextOM` on retext instances
 * Change exception-assertions in tests
 * Add nlcst-to-textom implementation to API
 * Add nlcst-to-textom as a dependency
 * Update parse-latin

0.2.0 / 2014-10-14
==================

 * Add retext-no-pants to desired plugins in Readme.md
 * Add retext-frequent-words to desired plugins in Readme.md
 * Add retext-emoticon to desired plugins in Readme.md
 * Refactor Readme.md

0.2.0-rc.5 / 2014-10-14
==================

 * Update textom
 * Add retext-walk to plugins in Readme.md
 * Add `bower_components/` to .npmignore
 * Move `benchmark/index.js` to `benchmark.js`

0.2.0-rc.4 / 2014-10-09
==================

 * Replace parse-english with NLCST in package keywords
 * Update .gitignore, .npmignore, bower ignore
 * Update textom, parse-latin for component, bower
 * Move spec to test.js
 * Refactor indent in .jscs.json
 * Remove browser test
 * Update textom, parse-latin
 * Add retext-inspect to plugins in Readme.md
 * Add retext-find to plugins in Readme.md
 * Add demo for retext-lancaster-stemmer
 * Add demo for retext-soundex

0.2.0-rc.3 / 2014-09-27
==================

 * Fix benchmarks failing benchmarks
 * Refactor Readme.md
 * Remove `npm update npm` from travis
 * Fix property order in manifest files
 * Update .gitignore, .npmignore
 * Update copyright in Readme.md
 * Remove testling
 * Update dependencies

0.2.0-rc.2 / 2014-09-21
==================

 * Merge branch 'feature/rename-apply-plugins-to-run'
 * Rename `applyPlugins` > `run`
 * Merge branch 'refactor/api'
 * Refactor API
 * Merge branch 'feature/fail-without-callback'
 * Add interface for error on omitted `done` for `parse`
 * Add spec for error on omitted `done` for `parse`
 * Merge branch 'feature/chainable-parse'
 * Add chainable `parse` and `applyPlugins`
 * Add spec for chainable `parse` and `applyPlugins`
 * Fix throwing interdependent plugins
 * Add spec for throwing interdependent plugins
 * Refactor spec
 * Refactor benchmark
 * Update benchmark results in docs
 * Update docs for async changes in 27dbfd7
 * Remove flashy demos from docs
 * Update and fix Installation in docs
 * Remove sauce labs from travis
 * Merge branch 'feature/async'

0.2.0-rc.1 / 2014-09-19
==================

 * Fix either istanbul or eslint failing in spec
 * Use ware to support async parsing
 * Add retext-syllable to Plugins

index.js
'use strict';

var nlcstToTextOM,
    TextOMConstructor,
    ParseLatin,
    Ware;

/*
 * Dependencies.
 */

nlcstToTextOM = require('nlcst-to-textom');
TextOMConstructor = require('textom');
ParseLatin = require('parse-latin');
Ware = require('ware');

/**
 * Construct an instance of `Retext`.
 *
 * @param {Function?} parser - the parser to use. Defaults
 *   to a new instance of `parse-latin`.
 * @constructor Retext
 */
function Retext(parser) {
    var self,
        TextOM;

    if (!parser) {
        parser = new ParseLatin();
    }

    self = this;
    TextOM = new TextOMConstructor();

    self.plugins = [];

    self.ware = new Ware();
    self.parser = parser;
    self.TextOM = TextOM;

    /*
     * Expose `TextOM` on `parser`, and vice versa.
     */

    parser.TextOM = TextOM;
    TextOM.parser = parser;
}

/**
 * Attaches `plugin`: a humble function.
 *
 * When `use` is called, the `plugin` is invoked with
 * the retext instance and an `options` object.
 * Code to initialize `plugin` should go here, such as
 * functionality to modify the object model (TextOM),
 * the parser (e.g., `parse-latin`), or the `retext`
 * instance itsekf.
 *
 * Optionally `plugin` can return a function which is
 * called every time the user invokes `parse` or `run`.
 * When that happends, that function is invoked with
 * a `Node` and an `options` object.
 * If `plugin` contains asynchronous functionality, it
 * should accept a third argument (`next`) and invoke
 * it on completion.
 *
 * @param {function(Retext, Object): function(Node, Object, Function?)} plugin
 * @return {Retext} - self
 */
Retext.prototype.use = function (plugin, options) {
    var self,
        onparse;

    if (typeof plugin !== 'function') {
        throw new TypeError(
            'Illegal invocation: `' + plugin + '` ' +
            'is not a valid argument for `Retext#use(plugin)`'
        );
    }

    if (typeof plugin.attach === 'function') {
        throw new TypeError(
            'Illegal invocation: `' + plugin + '` ' +
            'is not a valid argument for ' +
            '`Retext#use(plugin)`.\n' +
            'This breaking change, the removal of ' +
            '`attach`, occurred in 0.3.0-rc.2, see ' +
            'GitHub for more information.'
        );
    }

    self = this;

    /*
     * Ware does not know which plugins are attached,
     * only which `onrun` methods are. Thus, we have
     * a custom list of `plugins`, and here we check
     * against that.
     */

    if (self.plugins.indexOf(plugin) === -1) {
        self.plugins.push(plugin);

        onparse = plugin(self, options || {});

        if (typeof onparse === 'function') {
            self.ware.use(onparse);
        }
    }

    return self;
};

/**
 * Transform a given value into a node, applies attached
 * plugins to the node, and invokes `done` with either an
 * error (first argument) or the transformed node (second
 * argument).
 *
 * @param {string?} value - The value to transform.
 * @param {Object} [options={}] - Optional settings.
 * @param {function(Error, Node)} done - Callback to
 *   invoke when the transformations have completed.
 * @return {Retext} - self
 */
Retext.prototype.parse = function (value, options, done) {
    var self,
        nlcst;

    if (!done) {
        done = options;
        options = null;
    }

    if (typeof done !== 'function') {
        throw new TypeError(
            'Illegal invocation: `' + done + '` ' +
            'is not a valid argument for `Retext#parse(value, done)`.\n' +
            'This breaking change occurred in 0.2.0-rc.1, see GitHub for ' +
            'more information.'
        );
    }

    self = this;

    nlcst = self.parser.parse(value);

    self.run(nlcstToTextOM(self.TextOM, nlcst), options, done);

    return self;
};

/**
 * Applies attached plugins to `node` and invokes `done`
 * with either an error (first argument) or the transformed
 * `node` (second argument).
 *
 * @param {Node} node - The node to apply attached
 *   plugins to.
 * @param {Object} [options={}] - Optional settings.
 * @param {function(Error, Node)} done - Callback to
 *   invoke when the transformations have completed.
 * @return {Retext} - self
 */
Retext.prototype.run = function (node, options, done) {
    var self;

    if (!done) {
        done = options;
        options = null;
    }

    if (typeof done !== 'function') {
        throw new TypeError(
            'Illegal invocation: `' + done + '` ' +
            'is not a valid argument for ' +
            '`Retext#run(node, done)`.\n' +
            'This breaking change occurred in 0.2.0-rc.1, see GitHub for ' +
            'more information.'
        );
    }

    self = this;

    self.ware.run(node, options, done);

    return self;
};

/*
 * Expose `Retext`.
 */

module.exports = Retext;


benchmark.js
'use strict';

var Retext,
    sentence,
    paragraph,
    section,
    article;

/*
 * Module dependencies (retext).
 */

Retext = require('./');

/*
 * Test data.
 *
 * This includes:
 *
 * - An average sentence (w/ 20 words);
 * - An average paragraph (w/ 5 sentences);
 * - A (big?) section (w/ 10 paragraphs);
 * - A (big?) article (w/ 10 sections);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

article = section + Array(10).join('\n\n' + section);

/*
 * Benchmark suite.
 */

suite('retext.parse(value, callback);', function () {
    var retext;

    retext = new Retext();

    /*
     * Benchmark a paragraph.
     */

    bench('A paragraph (5 sentences, 100 words)',
        function (done) {
            retext.parse(paragraph, done);
        }
    );

    /*
     * Benchmark a section.
     */

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retext.parse(section, done);
        }
    );

    /*
     * Benchmark an article.
     */

    bench('An article (100 paragraphs, 500 sentences, 10,000 words)',
        function (done) {
            retext.parse(article, done);
        }
    );
});
